src/common/config.cc

```diff
@@ -726,8 +726,16 @@ void md_config_t::_apply_changes(std::ostream *oss)
 void md_config_t::call_all_observers()
 {
   std::map<md_config_obs_t*,std::set<std::string> > obs;
+  // Have the scope of the lock extend to the scope of
+  // handle_conf_change since that function expects to be called with
+  // the lock held. (And the comment in config.h says that is the
+  // expected behavior.)
+  //
+  // An alternative might be to pass a std::unique_lock to
+  // handle_conf_change and have a version of get_var that can take it
+  // by reference and lock as appropriate.
+  Mutex::Locker l(lock);
```

Comment:
I am late reviewing this, but this looks scary.  Holding md_config_t::lock while calling into handle_conf_change of observers: then the observers can't use get_val (it takes the lock) to react to the changes!

The comment in config.h is referring to the old struct-based config, where holding the lock would be a good thing because it would render it safe to read that struct, rather than in the get_val world where holding the lock is a recipe for deadlocks

Reply:
I did look through the rest of the file as a sanity check. It seems that we acquire the lock in apply_changes before calling _apply_changes which calls handle_conf_change.

So we had people observers being called with the lock held in one path and without the lock held in the other path. I wouldn't be against swinging the change in the other direction and releasing the lock before calling the observers from _apply_changes if you think that would be safer.

Reply:
We need the lock to cover the handle_conf_change methods until all of the strings are gone from legacy_config_opts.h.  Since injectargs etc already hold the lock this change seems correct.

The larger/more annoying issue is that the get_val() calls in handle_conf_change.  Maybe we can fix those just by marking the mutex recursive?

Reply:
That was why I added the _get_val template. A recursive mutex would work, too, as would passing a unique_lock reference into handle_conf_change that can be passed to get_val which can just look to see whether to acquire it or not.

Reply:
The unique_lock approach seems fragile: we'd have to update all current implementations that use get_val(), and going forward developers have to remember that all future calls to get_val() have to pass the unique_lock ref along to a get_val() variant that takes it.  And if they don't they'll only notice if they test changing the config option online and deadlock the config update thread.

My vote is recursive mutex, with a plan to remove it if/when legacy_config_opts ever gets emptied out...

Reply:
Huh. Looks like it already is recursive. I didn't realize that when I was worried about deadlock before. In that case all we really need to do is back out my _get_val change and just call get_val as normal.